### PR TITLE
Fb1275

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Source/Common/IntersectShapeVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
+#include "Ashkal/Ashkal/IntersectShapeVisitor.hpp"
+#include "Ashkal/Ashkal/ShapeVisitor.hpp"
+#include "Ashkal/Shapes/Circle.hpp"
+#include "Ashkal/Shapes/Line.hpp"
+#include "Ashkal/Shapes/Triangle.hpp"
+
+using namespace Ashkal;
+
+namespace {
+  bool intersects_line(const Point& p1, const Point& p2) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Remove this function, not used anywhere.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+  bool is_within_value(double pt1, double pt2, double value) {
+    auto high = std::max(pt1, pt2);
+    auto low = std::min(pt1, pt2);
+    return std::clamp(value, low, high) == value;
+  }
+
+  bool rect_contains(const std::array<std::pair<Point, Point>, 4> lines,
+      const Point& pt) {
+    if(is_within_value(lines[0].first.x, lines[1].first.x, pt.x) &&
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: ```return is_within_value(...) && ...;```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    return false;
+  }
+}
+
+bool Ashkal::intersects(const Square& square, const Shape& shape) {
+  struct Visitor final : ShapeVisitor {
+    bool operator()(const Square& square, const Shape& shape) {
+      m_intersects = false;
+      m_square = square;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Use a pointer:

```m_square = &square;```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      shape.accept(*this);
+      return m_intersects;
+    }
+
+    void visit(const Circle& circle) override {
+      auto lines = get_lines(m_square, invert(circle.get_transformation()));
+      for(auto& l : lines) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: The convention is:

```
for(auto& singular : plural) {
  ...
}
```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        if(distance(l.first, Point{0, 0}) <= 1) {
+          m_intersects = true;
+          return;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp:54]: Wow, this is good stuff man.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      for(auto& l : lines) {
+        auto s = slope(l.first, l.second);
+        auto a = 1 + (s * s);
+        auto y = y_intercept(l.first, s);
+        auto b = 2 * y * s;
+        auto c = -1 + (y * y);
+        auto root1 = (-b + std::sqrt(std::pow(b, 2) - (4 * a * c))) / (2 * a);
+        auto root2 = (-b - std::sqrt(std::pow(b, 2) - (4 * a * c))) / (2 * a);
+        if(s == 0) {
+          if(is_within_value(l.first.x, l.second.x, root1) ||
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]:
```
if(s == 0 && (is_within_value(...) || ...)) {
  ...
} else if(std::isnan(s) && ...) {
  ...
} else if(...) {
} else if(...) {
}

**Review by kranar — CHANGES_REQUESTED** [review]

File: Source/Common/IntersectShapeVisitor.cpp (hunk 6 above)
Comment (spireken) [reply Source/Common/IntersectShapeVisitor.cpp]: What difference does the else if() make here? If any of these if statements are true the the function returns from that point.

File: Include/Ashkal/Ashkal/IntersectShapeVisitor.hpp [hunk 7]
```diff
@@ -10,7 +10,7 @@ namespace Ashkal {
     \param square The Square to check for intersection with the Shape.
     \param shape The Shape to check for intersection with the Square.
   */
-  bool intersects(const Square& square, const Shape& shape);
+  bool intersects(Square& square, const Shape& shape);
```
Comment (kranar) [inline Include/Ashkal/Ashkal/IntersectShapeVisitor.hpp]: Pass by ```const&```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 8]
```diff
@@ -70,28 +63,22 @@ bool Ashkal::intersects(const Square& square, const Shape& shape) {
...
-        }
-        if(std::isnan(s)) {
-          if(is_within_value(l.first.y, l.second.y, root1) ||
-              is_within_value(l.first.y, l.second.y, root2)) {
-            m_intersects = true;
-            return;
-          }
-        }
-        if((is_within_value(l.first.x, l.second.x, root1) &&
+        if(s == 0 && is_within_value(l.first.x, l.second.x, root1) ||
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Correct this bug.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 9]
```diff
@@ -70,28 +63,22 @@ bool Ashkal::intersects(const Square& square, const Shape& shape) {
...
+            is_within_value(l.first.x, l.second.x, root2)) {
+          m_intersects = true;
+          return;
+        } else if(std::isnan(s) &&
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp:68]: Similar bug here.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 10]
```diff
@@ -70,28 +63,22 @@ bool Ashkal::intersects(const Square& square, const Shape& shape) {
...
-              is_within_value(l.first.y, l.second.y, root2)) {
-            m_intersects = true;
-            return;
-          }
-        }
-        if((is_within_value(l.first.x, l.second.x, root1) &&
+        if(s == 0 && is_within_value(l.first.x, l.second.x, root1) ||
+            is_within_value(l.first.x, l.second.x, root2)) {
+          m_intersects = true;
+          return;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: No need for this ```return;```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 11]
```diff
@@ -70,28 +63,22 @@ bool Ashkal::intersects(const Square& square, const Shape& shape) {
...
+        } else if(std::isnan(s) &&
+            is_within_value(l.first.y, l.second.y, root1) ||
+            is_within_value(l.first.y, l.second.y, root2)) {
+          m_intersects = true;
+          return;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Remove ```return;```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+  double y_intercept(const Point& p, double slope) {
+    if(std::isnan(slope)) {
+      return std::numeric_limits<double>::quiet_NaN();
+    }
+    return p.y - (p.x * slope);
+  }
+
+  bool is_within_value(double pt1, double pt2, double value) {
+    auto high = std::max(pt1, pt2);
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]:
```
auto ordered_pts = std::minmax(pt1, pt2);
return std::clamp(value, ordered_pts.first, ordered_pts.second);
```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    auto low = std::min(pt1, pt2);
+    return std::clamp(value, low, high) == value;
+  }
+
+  bool rect_contains(const std::array<std::pair<Point, Point>, 4> lines,
+      const Point& pt) {
+    return is_within_value(lines[0].first.x, lines[1].first.x, pt.x) &&
+        is_within_value(lines[0].first.y, lines[3].first.y, pt.y);
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Fix indentation.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    void visit(const Circle& circle) override {
+      auto lines = get_lines(*m_square, invert(circle.get_transformation()));
+      for(auto& line : lines) {
+        if(distance(line.first, Point{0, 0}) <= 1) {
+          m_intersects = true;
+          return;
+        }
+      }
+      for(auto& l : lines) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: ```auto& line : lines```

**Review by kranar — CHANGES_REQUESTED** [review]

**Pull request merged**